packages/dds/tree/src/feature-libraries/indexing/anchorTreeIndex.ts

```diff
@@ -0,0 +1,417 @@
+				yield filtered;
+			}
+		}
+	}
+
+	/**
+	 * Returns an iterable of key, value pairs for every entry in the index
+	 */
+	public *entries(): IterableIterator<[TKey, TValue]> {
+		for (const [key, nodes] of this.nodes.entries()) {
+			const filtered = this.getFilteredValue(nodes);
+			if (filtered !== undefined) {
+				yield [key, filtered];
+			}
+		}
+	}
+
+	public [Symbol.iterator](): IterableIterator<[TKey, TValue]> {
+		return this.entries();
+	}
+
+	/**
+	 * Applies the provided callback to each entry in the index.
+	 */
+	public forEach(
+		callbackfn: (value: TValue, key: TKey, map: AnchorTreeIndex<TKey, TValue>) => void,
+		thisArg?: unknown,
+	): void {
+		for (const [key, nodes] of this.nodes.entries()) {
+			const filtered = this.getFilteredValue(nodes);
+			if (filtered !== undefined) {
+				callbackfn.call(thisArg, filtered, key, this);
+			}
+		}
+	}
+
+	/**
+	 * Returns an iterable of key, value pairs for every entry in the index, including ones that are detached.
+	 * This function should only be used for testing purposes, it is not exposed as part of the public {@link TreeIndex} API.
+	 */
+	public *allEntries(): IterableIterator<[TKey, TValue]> {
+		for (const [key, nodes] of this.nodes.entries()) {
+			const value = this.getValue(nodes as unknown as TreeIndexNodes<AnchorNode>);
+			if (value !== undefined) {
+				yield [key, value];
+			}
+		}
+	}
+
+	/**
+	 * Disposes this index and all the anchors it holds onto.
+	 */
+	public [disposeSymbol](): void {
+		for (const anchors of this.anchors.values()) {
+			for (const anchor of anchors) {
+				this.forest.forgetAnchor(anchor);
+			}
+		}
+		this.anchors.clear();
+		Reflect.defineProperty(this, disposeSymbol, {
+			value: () => {
+				throw new Error("Index is already disposed");
+			},
+		});
+	}
+
+	/**
+	 * Given a cursor in node mode, indexes it.
+	 */
+	private indexNode(nodeCursor: ITreeSubscriptionCursor): void {
+		const keyFinder = getOrCreate(
+			this.keyFinders,
+			// the node schema type to look up
+			nodeCursor.type,
+			// if the indexer does not return a key finder for this schema, we cache a null value to indicate the indexer
+			// does not need to be called if this schema is encountered in the future
+			(schema) => this.indexer(schema) ?? null,
+		);
+
+		if (keyFinder !== null) {
+			const key = keyFinder(nodeCursor);
+			const anchor = nodeCursor.buildAnchor();
+			const anchorNode = this.forest.anchors.locate(anchor) ?? fail("expected anchor node");
+
+			// check if this anchor node already exists in the index
+			const existingKey = this.anchorKeys.get(anchorNode);
+			if (existingKey !== undefined) {
+				// if the node already exists but has the same key, we return early
+				if (existingKey === key) {
+					this.forest.forgetAnchor(anchor);
+					return;
+				} else {
+					// if the node has a different key, we remove the existing one first because it means the key had been detached
+					this.removeAnchor(anchorNode, existingKey);
```

Comment:
do we have a test case for this scenario? how can it happen? I thought we only remove things from the index when repair data is GC'd (and just filter until then)